tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11663,6 +11663,560 @@ static bool CheckBarrierCall(Sema &S, FunctionDecl *FD, CallExpr *CE,
   return false;
 }
 
+// MatVec Ops
+static const unsigned kMatVecMulOutputVectorIdx = 0;
+static const unsigned kMatVecMulOutputIsUnsignedIdx = 1;
+static const unsigned kMatVecMulInputVectorIdx = 2;
+static const unsigned kMatVecMulIsInputUnsignedIdx = 3;
+static const unsigned kMatVecMulInputInterpretationIdx = 4;
+// static const unsigned kMatVecMulMatrixBufferIdx = 5;
+// static const unsigned kMatVecMulMatrixOffsetIdx = 6;
+static const unsigned kMatVecMulMatrixInterpretationIdx = 7;
+static const unsigned kMatVecMulMatrixMIdx = 8;
+static const unsigned kMatVecMulMatrixKIdx = 9;
+static const unsigned kMatVecMulMatrixLayoutIdx = 10;
+static const unsigned kMatVecMulMatrixTransposeIdx = 11;
+static const unsigned kMatVecMulMatrixStrideIdx = 12;
+
+// MatVecAdd
+const unsigned kMatVecMulAddBiasInterpretation = 15;
+
+enum MatrixLayout {
+  MATRIX_LAYOUT_ROW_MAJOR = 0,
+  MATRIX_LAYOUT_COLUMN_MAJOR = 1,
+  MATRIX_LAYOUT_MUL_OPTIMAL = 2,
+  MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL = 3
+};
+
+bool IsValidMatrixLayoutForMulandMulAddOps(unsigned Layout) {
+  return Layout <= static_cast<unsigned>(
+                       MatrixLayout::MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL);
+}
+
+bool IsOptimalTypeMatrixLayout(unsigned Layout) {
+  return (Layout == (static_cast<unsigned>(
+                        MatrixLayout::MATRIX_LAYOUT_MUL_OPTIMAL)) ||
+          (Layout == (static_cast<unsigned>(
+                         MatrixLayout::MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL))));
+}
+
+bool IsValidTransposeForMatrixLayout(unsigned Layout, bool Transposed) {
+  switch (static_cast<MatrixLayout>(Layout)) {
+  case MatrixLayout::MATRIX_LAYOUT_ROW_MAJOR:
+  case MatrixLayout::MATRIX_LAYOUT_COLUMN_MAJOR:
+    return !Transposed;
+
+  default:
+    return true;
+  }
+}
+
+enum DataType {
```

Comment:
Is there a reason this can't use the dxilconstants.h enum?  There may well be a good one, just wanted to check.

Reply:
Yep this was a TBD, I wanted to check if we wanted to us the dxilconstants one or the linalg headers before updating. The linalg update will need a cmake change to add the path to include the headers.